uc-crux-llvm/src/UCCrux/LLVM/Specs/Apply.hs

```diff
@@ -0,0 +1,184 @@
+{-
+Module           : UCCrux.LLVM.Specs.Apply
+Description      : Apply collections of specifications for LLVM functions
+Copyright        : (c) Galois, Inc 2022
+License          : BSD3
+Maintainer       : Langston Barrett <langston@galois.com>
+Stability        : provisional
+
+TODO(lb): Improve tracking + logging of when specs failed to apply, and why.
+Should be easy enough given the information in 'CheckedConstraint'.
+-}
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+
+module UCCrux.LLVM.Specs.Apply
+  ( applySpecs
+  )
+where
+
+{- ORMOLU_DISABLE -}
+import           Prelude hiding (zip)
+
+import qualified Control.Lens as Lens
+import           Control.Monad.IO.Class (liftIO)
+import           Data.Maybe (fromMaybe)
+import           Data.Foldable (toList)
+import           Data.IORef (IORef)
+
+-- parameterized-utils
+import qualified Data.Parameterized.Context as Ctx
+
+-- what4
+import qualified What4.Interface as What4
+
+-- crucible
+import qualified Lang.Crucible.Backend as Crucible
+import           Lang.Crucible.Backend (IsSymInterface)
+import qualified Lang.Crucible.Simulator as Crucible
+
+-- crucible-llvm
+import           Lang.Crucible.LLVM.MemModel (HasLLVMAnn, Mem, MemImpl, MemOptions)
+import           Lang.Crucible.LLVM.TypeContext (TypeContext)
+
+-- crux-llvm
+import           Crux.LLVM.Overrides (ArchOk)
+
+-- uc-crux-llvm
+import           UCCrux.LLVM.Check (checkConstrainedShapes, SomeCheckedConstraint(..), CheckedConstraint(checkedPred))
+import           UCCrux.LLVM.Context.Module (ModuleContext)
+import           UCCrux.LLVM.ExprTracker (ExprTracker)
+import qualified UCCrux.LLVM.FullType.FuncSig as FS
+import           UCCrux.LLVM.FullType.Type (FullTypeRepr, MapToCrucibleType)
+import           UCCrux.LLVM.Module (FuncSymbol)
+import           UCCrux.LLVM.Postcondition.Apply (applyPostcond)
+import qualified UCCrux.LLVM.Postcondition.Type as Post
+import           UCCrux.LLVM.Specs.Type (Specs)
+import qualified UCCrux.LLVM.Specs.Type as Spec
+{- ORMOLU_ENABLE -}
+
+-- | Collapse all the preconditions of a 'Spec' to a single predicate
+matchPreconds ::
```

Comment:
> Why is it called matchPreconds? Are preconditions being matched? Against what?

My thinking was that they're being "matched" against the current state (i.e., arguments + LLVM memory). I'm certainly open to other names - I also describe it as "collapsing" in the Haddock, it could also be seen as a form of "checking" - although it returns the predicate, rather than asserting it...

> I guess reading ahead, this is meant to be a predicate in the head of an ite (or chain of symbolic branches) below?

That's how it's used in this branch, but like the "check" overrides, one could imagine checking preconditions for reasons other than applying the postcondition, as is done in the chain symbolic branches. I can add to the Haddock that this is used in `applySpecs` below.

Reply:
Added the above information in the Haddocks, let me know if it makes sense now!